### PR TITLE
[API] Add issue and comment attachments

### LINKED ISSUE #3690: Issue API and Attachments
- Gitea version (or commit ref): 1.3.3
- Git version: n/a
- Operating system:n/a
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [x ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [] Yes (provide example URL)
  - [ ] No
  - [ a] Not relevant
- Log gist:

## Description
The Issues API doesn't expose attachments, is this by design?
...

### PR DESCRIPTION
**Pull request by andrebruch**
The following API routes have been added:

Issues

- `GET /repos/{owner}/{repo}/issues/assets/{attachment_id}`
- `GET /repos/{owner}/{repo}/issues/{index}/assets`
- `POST /repos/{owner}/{repo}/issues/{index}/assets`
- `PATCH /repos/{owner}/{repo}/issues/assets/{attachment_id}`
- `DELETE /repos/{owner}/{repo}/issues/assets/{attachment_id}`

Comments

- `GET /repos/{owner}/{repo}/issues/comments/assets/{attachment_id}`
- `GET /repos/{owner}/{repo}/issues/comments/{id}/assets`
- `POST /repos/{owner}/{repo}/issues/comments/{id}/assets`
- `PATCH /repos/{owner}/{repo}/issues/comments/assets/{attachment_id}`
- `DELETE /repos/{owner}/{repo}/issues/comments/assets/{attachment_id}`

close #3690

### CODE REVIEW

**Comment by zeripath** [comment]
Please run `make generate-swagger` and commit the changes

File: modules/convert/issue.go [hunk 1]
```diff
@@ -26,21 +26,27 @@ func ToAPIIssue(issue *models.Issue) *api.Issue {
 		return &api.Issue{}
 	}
 
+	assets := make([]*api.Attachment, 0)
```
Comment (zeripath) [inline modules/convert/issue.go]: Add a third argument len(issue.Attachments) here

File: modules/convert/issue_comment.go [hunk 2]
```diff
@@ -11,14 +11,32 @@ import (
 
 // ToComment converts a models.Comment to the api.Comment format
 func ToComment(c *models.Comment) *api.Comment {
+	assets := make([]*api.Attachment, 0)
```
Comment (zeripath) [inline modules/convert/issue_comment.go]: Similar

File: routers/api/v1/repo/issue_attachment.go [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+
+	attachID := ctx.ParamsInt64(":asset")
+	attach, err := models.GetAttachmentByID(attachID)
+	if err != nil {
+		ctx.Error(http.StatusInternalServerError, "GetAttachmentByID", err)
+		return
+	}
+
+	if attach.IssueID == 0 {
+		log.Info("User requested attachment is not in issue, attachment_id: %v", attachID)
```
Comment (zeripath) [inline routers/api/v1/repo/issue_attachment.go]: Hmm not sure about this logging level. I suspect it's too high but I can see you don't want people to crawl it...

**Comment by andrebruch** [comment]
> Please run `make generate-swagger` and commit the changes

The swagger spec is already there.

**Comment by jolheiser** [comment]
Currently the linter is mad because it's detecting duplicate code.
https://drone.gitea.io/go-gitea/gitea/35748/1/4

**Comment by andrebruch** [comment]
The code is quite similar, but not the same.

File: routers/api/v1/repo/issue_attachment.go [hunk 4]
```diff
@@ -5,6 +5,7 @@
 package repo
 
 import (
+	issue_service "code.gitea.io/gitea/services/issue"
```
Comment (lunny) [inline routers/api/v1/repo/issue_attachment.go]: import order.

File: routers/repo/issue.go [hunk 5]
```diff
@@ -1673,17 +1673,17 @@ func UpdateIssueContent(ctx *context.Context) {
 		return
 	}
 
+	files := ctx.QueryStrings("files[]")
+	if err := updateAttachments(issue, files); err != nil {
+		ctx.ServerError("UpdateAttachments", err)
```
Comment (lunny) [inline routers/repo/issue.go]: return missed

**Comment by andrebruch** [comment]
It looks like a new line in the `v1_json.tmpl` file solved the duplicate code and `make generate-swagger` issues.

The `make generate-swagger` and `make swagger-check` commands remove the last line in the `v1_json.tmpl` file on my system.

**Comment by andrebruch** [comment]
The following test fails:
`testing-amd64` > `test-mssql` > `TestGit/HTTP/MergeFork/CreatePRAndMerge`

I just run the tests locally on a MSSQL database and there were no problems.

**Comment by 6543** [comment]
@andrebruch would be nice if you could resolve conflicts and adopt related changes :)

**Comment by zeripath** [comment]
I've fixed the conflicts but not looked at the comments yet - I think a lot of them may have been marked resolved by virtue of files being moved around.

**Comment by 6543** [comment]
only one thing missing: integration tests

File: routers/api/v1/api.go [hunk 6]
```diff
@@ -807,6 +807,13 @@ func Routes() *web.Route {
 								Get(repo.GetIssueCommentReactions).
 								Post(reqToken(), bind(api.EditReactionOption{}), repo.PostIssueCommentReaction).
 								Delete(reqToken(), bind(api.EditReactionOption{}), repo.DeleteIssueCommentReaction)
+							m.Combo("/assets").Get(repo.ListIssueCommentAttachments).
+								Post(reqToken(), reqRepoWriter(models.UnitTypeIssues), repo.CreateIssueCommentAttachment)
+						})
+						m.Group("/assets", func() {
+							m.Combo("/{asset}").Get(repo.GetIssueCommentAttachment).
+								Patch(reqToken(), reqRepoWriter(models.UnitTypeIssues), bind(api.EditAttachmentOptions{}), repo.EditIssueCommentAttachment).
+								Delete(reqToken(), reqRepoWriter(models.UnitTypeIssues), repo.DeleteIssueCommentAttachment)
```
Comment (zeripath) [inline routers/api/v1/api.go]: these permissions checks are incorrect.

File: routers/api/v1/api.go [hunk 7]
```diff
@@ -848,6 +855,15 @@ func Routes() *web.Route {
...
 							Delete(reqToken(), bind(api.EditReactionOption{}), repo.DeleteIssueReaction)
+						m.Group("/assets", func() {
+							m.Combo("").Get(repo.ListIssueAttachments).
+								Post(reqToken(), reqRepoWriter(models.UnitTypeIssues), repo.CreateIssueAttachment)
+						})
+					})
+					m.Group("/assets", func() {
+						m.Combo("/{asset}").Get(repo.GetIssueAttachment).
+							Patch(reqToken(), reqRepoWriter(models.UnitTypeIssues), bind(api.EditAttachmentOptions{}), repo.EditIssueAttachment).
+							Delete(reqToken(), reqRepoWriter(models.UnitTypeIssues), repo.DeleteIssueAttachment)
```
Comment (zeripath) [inline routers/api/v1/api.go]: similarly.

File: modules/convert/issue_comment.go [hunk 8]
```diff
@@ -11,14 +11,32 @@ import (
...
+		PRURL:       c.PRURL(),
+		Body:        c.Content,
+		Attachments: assets,
+		Created:     c.CreatedUnix.AsTime(),
+		Updated:     c.UpdatedUnix.AsTime(),
+	}
+}
+
+// ToCommentAttachment converts models.Attachment to api.Attachment
+func ToCommentAttachment(a *models.Attachment) *api.Attachment {
```
Comment (noerw) [inline modules/convert/issue_comment.go]: input + output types are equal to `ToIssueAttachment()`, so there is no future difference in behaviour to be expected.. I'd merge these functions.

File: routers/api/v1/repo/issue_attachment.go [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+		log.Debug("Requested attachment[%d] belongs to issue[%d, #%d] which is not in Repo: %-v.", attachID, issue.ID, issue.Index, ctx.Repo.Repository)
+		ctx.NotFound()
+		return nil
+	}
+	unitType := models.UnitTypeIssues
+	if issue.IsPull {
+		unitType = models.UnitTypePullRequests
+	}
+	if !ctx.IsUserRepoWriter([]models.UnitType{unitType}) && !ctx.IsUserRepoAdmin() && !ctx.IsUserSiteAdmin() {
+		ctx.Error(http.StatusForbidden, "reqRepoWriter", "user should have a permission to write to a repo")
```
Comment (noerw) [inline routers/api/v1/repo/issue_attachment.go]:
```suggestion
		ctx.Error(http.StatusForbidden, "reqIssueAttachment", "user should have a permission to write to a repo")
```

File: routers/api/v1/api.go (hunk 6 above)
Comment (noerw) [reply routers/api/v1/api.go]: Yeah, this is weird.
You apply two methods for checking permissions: Once here via middleware, but it seems you just copy pasted these, and then in certain handlers (eg `GetIssueAttachment()`) too.
I'd factor this in a way that you have an explicit permission check in middleware (you'll need at least two: reqReadAttachment, reqWriteAttachment), and no need for permission checks in the route handlers.

**Review by noerw — CHANGES_REQUESTED** [review]

**Comment by noerw** [comment]
@andrebruch oh, you're picking this up again? Yesterday I startet to rework this PR (rebase, fix permission issues, ...) If you want contact me on discord to coordinate :)

**Comment by noerw** [comment]
Here's my fork: https://github.com/go-gitea/gitea/compare/main...noerw:feature-add-issue-and-comment-attachments-api

**Comment by 6543** [comment]
I updated branch by merge & resolve conflicts, cherry-picked noerw commits & fix build

**Comment by 6543** [comment]
@noerw if you work on this - please use currend head as new base for your branch, else I'll get a hard time to cherry-pick things again ;)

**Comment by 6543** [comment]
```
+ make lint-backend
golangci-lint run --timeout 10m
routers/api/v1/repo/issue_comment_attachment.go:298:6: `getIssueCommentAttachmentSafeRead` is unused (deadcode)
func getIssueCommentAttachmentSafeRead(ctx *context.APIContext) *models.Attachment {
     ^
routers/api/v1/repo/issue_comment_attachment.go:311:6: `attachmentBelongsToRepoOrComment` is unused (deadcode)
func attachmentBelongsToRepoOrComment(ctx *context.APIContext, a *models.Attachment, issue *models.Issue, comment *models.Comment) (success bool) {
     ^
routers/api/v1/repo/issue_attachment.go:326: 326-348 lines are duplicate of `routers/api/v1/repo/issue_comment_attachment.go:311-333` (dupl)
func attachmentBelongsToRepoOrIssue(ctx *context.APIContext, a *models.Attachment, issue *models.Issue, comment *models.Comment) (success bool) {
	if a.RepoID != ctx.Repo.Repository.ID {
		log.Debug("Requested attachment[%d] does not belong to repo[%-v].", a.ID, ctx.Repo.Repository)
		ctx.NotFound()
		return
	}
	if a.IssueID == 0 {
		// catch people trying to get release assets ;)
		log.Debug("Requested attachment[%d] is not in an issue.", a.ID)
		ctx.NotFound()
		return
	} else if issue != nil && a.IssueID != issue.ID {
		log.Debug("Requested attachment[%d] does not belong to issue[%d, #%d].", a.ID, issue.ID, issue.Index)
		ctx.NotFound()
		return
	}
	if comment != nil && a.CommentID != comment.ID {
		log.Debug("Requested attachment[%d] does not belong to comment[%d].", a.ID, comment.ID)
		ctx.NotFound()
		return
	}
	return true
}
make: *** [Makefile:770: golangci-lint] Error 1
```

**Comment by andrebruch** [comment]
As just discussed on Discord, I'm taking care of the integration tests :)

**Comment by noerw** [comment]
@andrebruch @6543 I added a couple of commits, feel free to cherry pick them:
https://github.com/andrebruch/gitea/compare/feature-add-issue-and-comment-attachments-api...noerw:feature-add-issue-and-comment-attachments-api

- finalized the rework of access permissions
- moved all `/assets/{assetID}` routes under the namespace of the respective issue or comment ID
- fixed OpenAPI docs a bit

File: routers/web/repo/issue.go [hunk 10]
```diff
@@ -1730,12 +1730,19 @@ func UpdateIssueContent(ctx *context.Context) {
...
 			ctx.ServerError("UpdateAttachments", err)
 			return
 		}
 	}
 
+	content := ctx.FormString("content")
+	if err := issue_service.ChangeContent(issue, ctx.User, content); err != nil {
+		ctx.ServerError("ChangeContent", err)
+		return
+	}
```
Comment (noerw) [inline routers/web/repo/issue.go:1863]: I'm unsure what this change is about, could this be erroneously introduced while merging current master?

File: modules/notification/webhook/webhook.go [hunk 11]
```diff
@@ -310,8 +310,11 @@ func (m *webhookNotifier) NotifyNewPullRequest(pull *models.PullRequest, mention
 }
 
 func (m *webhookNotifier) NotifyIssueChangeContent(doer *models.User, issue *models.Issue, oldContent string) {
-	mode, _ := models.AccessLevel(issue.Poster, issue.Repo)
 	var err error
+	if err = issue.LoadRepo(); err != nil {
+		log.Error("NotifyIssueChangeContent: coulnd't load repo", err)
```
Comment (lunny) [inline modules/notification/webhook/webhook.go:341]: Then just return, otherwise it will panic.

**Comment by lunny** [comment]
Per https://github.com/go-gitea/gitea/pull/14601/files#r725277516 and https://github.com/go-gitea/gitea/pull/14601/files#r748950335, otherwise LGTM.

File: modules/notification/webhook/webhook.go (hunk 11 above)
Comment (tomasandrle) [reply modules/notification/webhook/webhook.go:341]: Typo: coulnd't

**Comment by noerw** [comment]
There's still one commit that could be cherry picked from my branch: https://github.com/andrebruch/gitea/compare/feature-add-issue-and-comment-attachments-api...noerw:feature-add-issue-and-comment-attachments-api
This fixes empty or non-json error messages being returned

**Comment by mattrpav** [comment]
This one looks close.. is there a chance this makes it back into 1.16.0? We are preparing a migration from GitHub + JIRA to Gitea for engineering tasks. We plan on open sourcing the JIRA migration tool for others, since JIRA is EOL self-hosting option.

**Comment by singuliere** [comment]
@andrebruch could you please resolve the conflicts?

**Comment by mattrpav** [comment]
Checking in-- how's this looking for inclusion with a 1.17.0 or a 1.16.x release?

**Comment by 6543** [comment]
1.17.0 ...

**Comment by 6543** [comment]
... I can have a look at finishing it

File: modules/context/api.go [hunk 12]
```diff
@@ -109,17 +109,18 @@ func (ctx *APIContext) Error(status int, title string, obj interface{}) {
 // InternalServerError responds with an error message to the client with the error as a message
 // and the file and line of the caller.
 func (ctx *APIContext) InternalServerError(err error) {
-	log.ErrorWithSkip(1, "InternalServerError: %v", err)
+	ctx.Error(http.StatusInternalServerError, "InternalServerError", err)
```
Comment (lunny) [inline modules/context/api.go]: Why change these functions? It seems they are break changes?

**Review by lunny — CHANGES_REQUESTED** [review]
Please follow my comments.

File: routers/api/v1/repo/issue_comment_attachment.go [hunk 13]
```diff
@@ -0,0 +1,389 @@
...
+		return nil
+	}
+	if !attachmentBelongsToRepoOrComment(ctx, attach, comment) {
+		return nil
+	}
+	return attach
+}
+
+func canUserWriteIssueCommentAttachment(ctx *context.APIContext, c *models.Comment) (success bool) {
+	canEditComment := ctx.Doer.ID == c.PosterID || ctx.IsUserRepoAdmin() || ctx.IsUserSiteAdmin()
```
Comment (lunny) [inline routers/api/v1/repo/issue_comment_attachment.go:362]: We should check if user has a write permission to the issue.

**Comment by mattrpav** [comment]
Friendly comment to bubble up. Thanks all!

**Comment by mattrpav** [comment]
Hi @6543 @lunny checking in. Any update to the plans for this feature request?

**Comment by lunny** [comment]
The review comments should be followed and conflict files should be resolved.

**Comment by 6543** [comment]
original contributor is not responding so everybody looking at this pull and want it ...

if you have time checkout to the pull-head and create a new branch on you own fork and submit a pull ... then fix all open mentioned issues and we will merge that one :)

**Comment by mattrpav** [comment]
Hi @6543 @lunny @andrebruch -  Thanks for the follow-ups. I'm not a Go programmer, so I don't think anyone wants to see me tackle the patch at this point ;-). Is there a tip jar or other option to financially back this feature?

I have committed to writing a JIRA-to-Gitea migration tool and publish blog for others to benefit-- and we are happy to kick in money for a Gitea/Go programmer to finish out this PR.

Also, this feature will be needed by any organization that wants to use Gitea and maintain automated security compliance (SOC2, ISO27001, etc).

Thanks!

**Comment by lunny** [comment]
> Hi @6543 @lunny @andrebruch - Thanks for the follow-ups. I'm not a Go programmer, so I don't think anyone wants to see me tackle the patch at this point ;-). Is there a tip jar or other option to financially back this feature?
> 
> I have committed to writing a JIRA-to-Gitea migration tool and publish blog for others to benefit-- and we are happy to kick in money for a Gitea/Go programmer to finish out this PR.
> 
> Also, this feature will be needed by any organization that wants to use Gitea and maintain automated security compliance (SOC2, ISO27001, etc).
> 
> Thanks!

Thank you very much! You can contribute in bountysource for #3690 but even that I cannot guarantee this could be resolved in a short days.

**Comment by mattrpav** [comment]
@lunny I just put $100 (USD). How much $ do you think to have it completed by December 31st?

**Comment by lunny** [comment]
Let's continue in #21783

**Pull request closed without merging**